presto-array/src/main/java/com/facebook/presto/array/ObjectBigArray.java

```diff
@@ -72,6 +80,24 @@ public T get(long index)
         return (T) array[segment(index)][offset(index)];
     }
 
+    /**
+     * Gets the element of this big array at specified index and then sets the provided
+     * replacement value to the same index
+     *
+     * @param index a position in this big array.
+     *
+     * @return the previously stored value for the specified index
+     */
+    @SuppressWarnings("unchecked")
+    public T getAndSet(long index, T replacement)
+    {
+        Object[] segment = array[segment(index)];
+        int offset = offset(index);
+        T result = (T) segment[offset];
+        segment[offset] = replacement;
+        return result;
+    }
```

Comment:
How much perf improvement this will bring comparing using `get()` and `set()`

Reply:
It’s hard to quantify in isolation, but in general the access pattern (either reads or writes) on BigArray are opaque to the JVM (for the purposes of bounds check elimination) and to the CPU in being able to speculate through the 2D array for the memory accesses.

This doesn’t matter that much when everything fits in the CPU cache, but when accesses start hitting main memory, the extent to which the CPU can speculate beyond the access to the subsequent operations in the pipeline is critical. Filling the pipeline with more data-dependent accesses than necessary hurts performance for that reason, which is why changes like this and like https://github.com/trinodb/trino/pull/6591 are effective.